tests/test_utils.py

```diff
@@ -824,3 +826,118 @@ def test_non_operations_in_list(self, arg):
             ValueError, match="The given operation_list does not only contain Operations"
         ):
             pu.inv(arg)
+
+
+class TestSpy:
+    """Tests the Spy context."""
+
+    def test_single_output(self, ):
+        """Tests a qnode with a 1-d output."""
+
+        qml.enable_tape()
+
+        dev = qml.device('default.qubit', wires=1)
+
+        @qml.qnode(dev)
+        def circuit(w):
+            qml.RX(w, wires=0)
+            qml.Hadamard(wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        with Spy() as counter:
+            # 2 executions
+            circuit(0.1)
+            circuit(0.5)
+
+            # 3 executions
+            g = qml.grad(circuit)
+            g(0.1)
+
+            # 1 execution
+            circuit.qtape.execute(dev)
+
+        assert counter.counts == 6
+
+        qml.disable_tape()
+
+    def test_multi_output(self):
+        """Tests a qnode with a 2-d output."""
```

Comment:
Some other edge cases that are worth testing:

* Measurements that return arrays (`qml.probs` or `qml.state`)

* Multiple arrays (`return qml.probs(wires=0), qml.probs(wires=1)`)

* Ragged arrays (`return qml.probs(wires=0), qml.expval(qml.PauliZ(1))`

* Samples (`return qml.sample(wires=0)`)